### ISSUE TITLE
bug(query): document updates to nested structures overwrite subcollections

### ISSUE THREAD

**Issue by mechanical-turk** [prescottprue/redux-firestore#9]
Let's say we have the following schema:

```
{
   title: String,
   completed: Boolean,
   todo_items: SubCollection
}
```

Let's add the following `firestoreConnect(['/todos'])`

This will set `store.firestore.data.todos` to something like the following:
```js
{
  'askldaskdadm123': {
    title: 'First todos list',
  },
  'omfnerfjknn103': {
    title: 'Second todos list',
  },
}
```
Now let's add the following: `firestoreConnect(['/todos/6mjAkNg3XvB3v55LDIjy/todo_items'])`
This will add the `todo_items` key to every todos list under `store.firestore.data.todos`, making it look like the following:

```
{
  'askldaskdadm123': {
    title: 'First todos list',
    todo_items: {
      'k3j2n42304': {
        title: 'first item under first list',
        completed: false,
      },
    },
  },
  'omfnerfjknn103': {
    title: 'Second todos list',
    todo_items: .....
  },
}
```

This is extremely convenient, but the data-structure will break when we update a todos list on the document level.

If we update the title of the second todos list to `Something else`, it will lose its todo_items, making `store.firestore.data.todos` look like the following:

```
{
  'askldaskdadm123': {
    title: 'First todos list',
    todo_items: {
      'k3j2n42304': {
        title: 'first item under first list',
        completed: false,
      },
    },
  },
  'omfnerfjknn103': {
    title: 'Something else',
  },
}
```

This is probably a reducer bug. When the snapshot of the todos list with an id: `todosListId` is received, I believe the reducer will automatically set `store.firestore.todos[todosListId]` to the value of that todos list, which will effectively overwrite the subcollections.

Instead, maybe we should first check if there's already some data in there, and then use the spread operator to only update with the new data.

**Comment by prescottprue** [comment]
Yup, I agree with your assessment. The same thing happened in the reducer of `react-redux-firebase` with real time DB. There were some that found it helpful and others found it to be unexpected, so in v2 there is a new action called `MERGE`.

Not sure if the naming still fits, but maybe something similar for this?

We might want to make this an option that can be passed with the query (that was actually a feature request in `react-redux-firebase`), something like `merge: Boolean`.

In the long term future it is the hope to expose the way data is handled in certain situations as config options so that they can easily be overridden/modified.

**Comment by prescottprue** [comment]
@mechanical-turk Can you give the v0.1.0-alpha.2 release a try? I figured out some better handling of collection stuff to support the fix for #12.

**Comment by prescottprue** [comment]
Should also be fixed by v0.1.0-beta.5. Let me know if it isn't working as expected.
